### **카이제곱 검정**
#### **적합도 검정**
##### **카이제곱 적합도 검정**
- 1개의 범주형 변수가 특정 분포를 잘 따르고 있는지 검정.


`stats.chisquare(observed, expected, ddof, axis)`
- `observed` : 관측된 빈도 리스트
- `expected` : 기대 빈도 리스트
- `ddof` : 자유도 조정
- `axis` : 축


Q. 어떤 도시에서 300명을 대상으로 아이스크림 맛 선호도를 조사한 결과, 바닐라 150명, 초코 120명, 딸기 30명이었다.


전국적인 아이스크림 맛 선호도 조사 결과를 통해 알려진 비율은 바닐라 50%, 초코 35%, 딸기 15%로 알려져 있다.


이 도시의 아이스크림 맛 선호도는 전국적인 맛 선호도와 차이가 있는가? (유의수준 0.05)

In [6]:
from scipy import stats

observed = [150, 120, 30]
expected = [0.5*300, 0.35*300, 0.15*300]

stats.chisquare(observed, expected)

Power_divergenceResult(statistic=7.142857142857142, pvalue=0.028115659748972056)

H0 : 이 도시의 아이스크림 맛 선호도는 전국적인 맛 선호도와 차이가 없다.


H1 : 이 도시의 아이스크림 맛 선호도는 전국적인 맛 선호도와 차이가 있다.

**p-value가 0.05보다 작으므로 대립가설을 채택한다.**

---

#### **독립성 검정**
##### **교차표 기반 카이제곱 검정**
`stats.chi2_contingency(table, correction=True)`
- correction : 연속성 보정 여부 


Q. 성별에 따라 운동을 좋아하는지 조사한 결과다. 성별과 운동 선호도가 독립적인지 가설 검정을 실시하시오. (유의수준 0.05)

남자 : 좋아함 80명, 좋아하지 않음 30명


여자 : 좋아함 90명, 좋아하지 않음 10명

- H0 : 성별과 운동 선호도는 독립적이다.
- H1 : 성별과 운동 선호도는 독립적이지 않다.

풀이 1. 교차표 생성

|       | 좋아함 | 좋아하지 않음 |
|-------|--------|------------|
| 남자  | 80     | 30          |
| 여자  | 90     | 10          |


In [15]:
# 교차표 생성
import pandas as pd

df = pd.DataFrame({'좋아함':[80, 90],
                   '좋아하지 않음':[30, 10]},
                   index=['남자','여자'])
df

,좋아함,좋아하지 않음
남자,80,30
여자,90,10


In [16]:
# 독립성 검정
from scipy import stats

stats.chi2_contingency(df)

Chi2ContingencyResult(statistic=9.045792112299468, pvalue=0.0026330012530379632, dof=1, expected_freq=array([[89.04761905, 20.95238095],
       [80.95238095, 19.04761905]]))

풀이 2. raw 데이터

In [35]:
import pandas as pd
data = {
    '성별': ['남자']*110 + ['여자']*100,
    '운동': ['좋아함']*80 + ['좋아하지 않음']*30 + ['좋아함']*90 + ['좋아하지 않음']*10
}
df = pd.DataFrame(data)
df

,성별,운동
0,남자,좋아함
1,남자,좋아함
2,남자,좋아함
3,남자,좋아함
4,남자,좋아함
...,...,...
205,여자,좋아하지 않음
206,여자,좋아하지 않음
207,여자,좋아하지 않음
208,여자,좋아하지 않음


In [38]:
# 크로스탭을 활용해 교차표로 변경
df = pd.crosstab(df['성별'], df['운동'])
df

운동,좋아하지 않음,좋아함
성별,,
남자,30,80
여자,10,90


In [40]:
from scipy import stats

stats.chi2_contingency(df)

Chi2ContingencyResult(statistic=9.045792112299468, pvalue=0.0026330012530379632, dof=1, expected_freq=array([[20.95238095, 89.04761905],
       [19.04761905, 80.95238095]]))

`expected_freq=array([[20.95238095, 89.04761905], [19.04761905, 80.95238095]])`

- 배열 첫번째 값 : 남자가 좋아하지 않음을 선택할 것으로 기대되는 수
- 배열 두번째 값 : 남자가 좋아함을 선택할 것으로 기대되는 수
- 배열 세번째 값 : 여자가 좋아하지 않음을 선택할 것으로 기대되는 수
- 배열 네번째 값 : 여자가 좋아함을 선택할 것으로 기대되는 수

---
#### **동질성 검정**
2개 이상의 집단에서 분산의 동질성을 가졌는지 검정

ex) 2개의 다른 학교에서 학생들의 운동 선호도가 동일한가?


- H0 : 모든 그룹의 분포나 비율은 동일하다.
- H1 : 각 그룹의 분포나 비율은 동일하지 않다.

##### **교차표 기반 카이제곱 검정**
`stats.chi2_contingency(table, correction=True)`

In [43]:
import pandas as pd
from scipy import stats

df = pd.DataFrame([[50, 50], [30, 70]])
stats.chi2_contingency(df)


Chi2ContingencyResult(statistic=7.520833333333334, pvalue=0.006098945931214352, dof=1, expected_freq=array([[40., 60.],
       [40., 60.]]))

In [49]:
import pandas as pd
data = {
    '학과': ['통계학과']*100 + ['컴퓨터공학과']*100,
    '동아리가입여부': ['가입']*50 + ['미가입']*50 + ['가입']*30 + ['미가입']*70
}
df = pd.DataFrame(data)
df.sample(5)

,학과,동아리가입여부
114,컴퓨터공학과,가입
72,통계학과,미가입
17,통계학과,가입
64,통계학과,미가입
197,컴퓨터공학과,미가입


In [50]:
df = pd.crosstab(df['학과'], df['동아리가입여부'])
df

동아리가입여부,가입,미가입
학과,,
컴퓨터공학과,30,70
통계학과,50,50


In [51]:
stats.chi2_contingency(df)

Chi2ContingencyResult(statistic=7.520833333333334, pvalue=0.006098945931214352, dof=1, expected_freq=array([[40., 60.],
       [40., 60.]]))